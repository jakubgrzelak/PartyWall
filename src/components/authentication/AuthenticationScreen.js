import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import {
  View,
  Text,
  KeyboardAvoidingView,
  TextInput,
  StyleSheet,
  Platform,
  Dimensions,
  Keyboard,
  TouchableOpacity,
} from 'react-native';
import { Formik } from 'formik';
import { inputStyle } from '../../helpers/styleHelper';
import { Button } from '../buttons/Button';
import { commonStyles } from '../../themes/common';
import {
  createAccountValidationSchema,
  signinValidationSchema,
} from '../../validators/formValidators';
// import { loginSuccess, signupRequest } from '../../redux/user/actions';
import Colors from '../../themes/colors';
import { Header } from '../common/Header';
import { BaseText } from '../common/BaseText';

const SCREEN_HEIGHT = Dimensions.get('window').height;

export const AuthenticationScreen = ({ navigation, createAccount = false }) => {
  const dispatch = useDispatch();
  const [errorMessaage, setErrorMessage] = useState(null);

  const description = createAccount
    ? 'Please enter your user name and password in order to create new account'
    : 'Please enter your credentials in order to login';
  const submitButtonText = createAccount ? 'Sign Up' : 'Login';
  const bottomLinkDescription = createAccount ? 'If you have an account' : 'If you do not have an account';
  const bottomLinkText = createAccount ? 'please login here!' : 'please create one here!';

  function handleLoginPress(values) {
    Keyboard.dismiss();
    // dispatch(signupRequest(values));
  }

  function handleCreateAccount(values) {
    setErrorMessage(null);
    Keyboard.dismiss();
    // createUser({ variables: { ...values } });
  }

  function onBottomLinkPress() {
    if (createAccount) {
      navigation.navigate('LoginScreen')
    } else {
      navigation.navigate('CreateAccountScreen')
    }
  }

  return (
    <View style={{ height: SCREEN_HEIGHT - 100 }}>
      <KeyboardAvoidingView
        behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
        style={styles.container}
        keyboardShouldPersistTaps={true}>
        {/* {error && <Text style={commonStyles.errorMessage}>{error}</Text>} */}
        {/* <Loader loading={isLoading} /> */}
        <BaseText style={[commonStyles.descriptionText, styles.headerText]}>
          {description}
        </BaseText>
        {errorMessaage && (
          <BaseText style={commonStyles.errorMessage}>{errorMessaage}</BaseText>
        )}
        <Formik
          initialValues={{
            password: '',
            nickname: '',
            passwordConfirmation: '',
          }}
          onSubmit={createAccount ? handleCreateAccount : handleLoginPress}
          validationSchema={
            createAccount
              ? createAccountValidationSchema
              : signinValidationSchema
          }>
          {({
            handleChange,
            handleSubmit,
            handleBlur,
            values,
            errors,
            touched,
          }) => (
            <View style={styles.content}>
              {errors.password && touched.password && (
                <Text style={commonStyles.errorMessage}>{errors.password}</Text>
              )}
              <TextInput
                onChangeText={handleChange('nickname')}
                onBlur={handleBlur('nickname')}
                value={values.nickname}
                placeholder="Nickname"
                style={inputStyle(
                  errors,
                  touched,
                  styles.inputStyle,
                  'nickname',
                )}
                autoCapitalize="none"
              />
              {errors.nickname && touched.nickname && (
                <Text style={commonStyles.errorMessage}>{errors.nickname}</Text>
              )}
              <TextInput
                onChangeText={handleChange('password')}
                onBlur={handleBlur('password')}
                value={values.password}
                secureTextEntry={true}
                placeholder="Password"
                style={inputStyle(
                  errors,
                  touched,
                  styles.inputStyle,
                  'password',
                )}
                autoCapitalize="none"
              />
              {errors.password && touched.password && (
                <Text style={commonStyles.errorMessage}>{errors.password}</Text>
              )}
              {createAccount && (
                <TextInput
                  onChangeText={handleChange('passwordConfirmation')}
                  onBlur={handleBlur('passwordConfirmation')}
                  value={values.passwordConfirmation}
                  secureTextEntry={true}
                  placeholder="Password Confirmation"
                  style={inputStyle(
                    errors,
                    touched,
                    styles.inputStyle,
                    'passwordConfirmation',
                  )}
                  autoCapitalize="none"
                />
              )}
              {errors.passwordConfirmation && touched.passwordConfirmation && (
                <Text style={commonStyles.errorMessage}>
                  {errors.passwordConfirmation}
                </Text>
              )}
              <Button
                onPress={handleSubmit}
                buttonStyles={styles.loginButton}
                text={submitButtonText}
              />
            </View>
          )}
        </Formik>
      </KeyboardAvoidingView>
      <View style={styles.bottomTextContainer}>
        <BaseText style={styles.text}>
          {bottomLinkDescription}
          <TouchableOpacity onPress={onBottomLinkPress}>
              <BaseText style={styles.linkText}>{bottomLinkText}</BaseText>
          </TouchableOpacity>
        </BaseText>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  content: {
    width: '100%',
    alignItems: 'center',
  },
  loginButton: {
    marginTop: 10,
    width: 200,
    alignSelf: 'center',
  },
  inputStyle: {
    ...commonStyles.textInputStyle,
    marginVertical: 5,
    width: '70%',
    marginBottom: 10,
    flexDirection: 'row',
  },
  headerText: {
    textAlign: 'center',
    paddingBottom: 20,
    paddingHorizontal: '15%',
  },
  textButton: {
    paddingTop: 10,
    paddingBottom: 20,
  },
  buttonText: {
    fontSize: 15,
    color: Colors.activeDot,
  },
  phoneCodeContent: {
    justifyContent: 'center',
    height: '100%',
    alignSelf: 'flex-start',
    marginRight: 5,
  },
  phoneInputContainer: {
    ...commonStyles.textInputStyle,
    marginVertical: 5,
    width: '70%',
    marginBottom: 10,
    alignItems: 'center',
    justifyContent: 'center',
  },
  phoneText: {
    fontSize: 15,
    fontWeight: '500',
    color: Colors.lightText,
  },
  phoneInput: {
    width: '70%',
  },
  bottomTextContainer: {
    position: 'absolute',
    width: '100%',
    bottom: 20,
    justifyContent: 'center',
    alignItems: 'center',
  },
  text: {
    fontSize: 16,
    width: '60%',
    textAlign: 'center',
  },
  linkText: {
    color: Colors.linkText,
    fontWeight: '500',
    textAlign: 'center',
    fontSize: 16,
  },
});
