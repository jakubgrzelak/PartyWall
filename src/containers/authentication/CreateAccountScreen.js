import React from 'react';
import { SafeAreaView } from 'react-native-safe-area-context';
import { AuthenticationScreen } from '../../components/authentication/AuthenticationScreen';

export const CreateAccountScreen = ({ navigation }) => (
  <SafeAreaView>
    <AuthenticationScreen navigation={navigation} createAccount />
  </SafeAreaView>
);
