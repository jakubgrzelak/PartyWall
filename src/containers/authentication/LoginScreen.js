import React from 'react';
import { SafeAreaView } from 'react-native-safe-area-context';
import { AuthenticationScreen } from '../../components/authentication/AuthenticationScreen';

export const LoginScreen = ({ navigation }) => (
  <SafeAreaView>
    <AuthenticationScreen navigation={navigation} />
  </SafeAreaView>
);

