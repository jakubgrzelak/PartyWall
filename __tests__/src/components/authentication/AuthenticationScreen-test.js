import React from 'react';
import { render, fireEvent, act, waitFor } from '@testing-library/react-native';
import { AuthenticationScreen } from '../../../../src/components/authentication/AuthenticationScreen';
import { useSelector } from "react-redux";

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

beforeEach(() => {
  useSelector.mockImplementation(callback => {
    return callback({ user: { isLoading: false, error: null }});
  });
});

afterEach(() => {
  useSelector.mockClear();
});

describe('Rendering', () => {
  it('should match to snapshot - login', () => {
    const tree = render(<AuthenticationScreen />).toJSON();
    expect(tree).toMatchSnapshot()
  });

  it('should match to snapshot - creating account', () => {
    const tree = render(<AuthenticationScreen createAccount />).toJSON();
    expect(tree).toMatchSnapshot()
  });
});

describe('Change input - login', () => {
  it('should change values', async () => {
    const { getByPlaceholderText } = render(<AuthenticationScreen />);
    const usernameInput = getByPlaceholderText('username');
    const passwordInput = getByPlaceholderText('password');

    fireEvent.changeText(usernameInput, 'abc');
    fireEvent.changeText(passwordInput, 'asd');

    await waitFor(() => expect(usernameInput.props.value).toEqual('abc'));
  });

  it('should render error messages', async () => {
    const { getByText  } = render(<AuthenticationScreen />);
    const loginButton = getByText('Login');
    fireEvent.press(loginButton);
    await waitFor(() => expect(getByText('Username is required')).not.toBeNull());
    await waitFor(() => expect(getByText('Password is required')).not.toBeNull());
  });
});

describe('Change input - signup', () => {
  it('should change values correctly', async () => {
    const { getByPlaceholderText, getByText } = render(<AuthenticationScreen createAccount />);
    const usernameInput = getByPlaceholderText('username');
    const passwordInput = getByPlaceholderText('password');
    const passwordConfirmation = getByPlaceholderText('Password Confirmation');

    fireEvent.changeText(usernameInput, 'abc');
    fireEvent.changeText(passwordInput, 'asd');
    fireEvent.changeText(passwordConfirmation, 'asd');

    await waitFor(() => expect(usernameInput.props.value).toEqual('abc'));
    await waitFor(() => expect(passwordInput.props.value).toEqual('asd'));
    await waitFor(() => expect(passwordConfirmation.props.value).toEqual('asd'));
  });

  it('should render error messages - passwords not match', async () => {
    const { getByText, getByPlaceholderText } = render(<AuthenticationScreen createAccount />);
    const passwordInput = getByPlaceholderText('password');
    const passwordConfirmation = getByPlaceholderText('Password Confirmation');
    const signupButton = getByText('Sign Up');

    fireEvent.changeText(passwordInput, 'asd');
    fireEvent.changeText(passwordConfirmation, 'dsd');

    fireEvent.press(signupButton);
    
    await waitFor(() => expect(getByText('Passwords must match')).not.toBeNull());
  });
});

