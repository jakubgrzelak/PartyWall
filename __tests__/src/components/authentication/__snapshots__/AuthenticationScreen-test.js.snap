// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering should match to snapshot - creating account 1`] = `
<View
  style={
    Object {
      "height": 1234,
    }
  }
>
  <View
    keyboardShouldPersistTaps={true}
    onLayout={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <Modal
      animationType="none"
      hardwareAccelerated={false}
      transparent={true}
      visible={false}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#00000040",
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "space-around",
            "zIndex": 0,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "borderRadius": 10,
              "display": "flex",
              "height": 100,
              "justifyContent": "space-around",
              "width": 100,
              "zIndex": 0,
            }
          }
        >
          <ActivityIndicator
            animating={false}
            color="#157dfb"
            hidesWhenStopped={true}
            size="large"
          />
        </View>
      </View>
    </Modal>
    <Text
      style={
        Array [
          Object {
            "fontWeight": "300",
          },
          Array [
            Object {
              "color": "#808080",
              "fontSize": 18,
              "paddingBottom": 35,
              "paddingTop": 15,
            },
            Object {
              "paddingBottom": 20,
              "paddingHorizontal": "15%",
              "textAlign": "center",
            },
          ],
        ]
      }
    >
      Please enter your user name and password in order to create new account
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "width": "100%",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholder="username"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
              "borderColor": "gray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#808080",
              "fontSize": 15,
              "fontWeight": "500",
              "height": 50,
              "paddingHorizontal": 10,
            },
            Object {
              "backgroundColor": "#fff",
              "borderColor": "gray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#808080",
              "flexDirection": "row",
              "fontSize": 15,
              "fontWeight": "500",
              "height": 50,
              "marginBottom": 10,
              "marginVertical": 5,
              "paddingHorizontal": 10,
              "width": "70%",
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholder="password"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
              "borderColor": "gray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#808080",
              "fontSize": 15,
              "fontWeight": "500",
              "height": 50,
              "paddingHorizontal": 10,
            },
            Object {
              "backgroundColor": "#fff",
              "borderColor": "gray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#808080",
              "flexDirection": "row",
              "fontSize": 15,
              "fontWeight": "500",
              "height": 50,
              "marginBottom": 10,
              "marginVertical": 5,
              "paddingHorizontal": 10,
              "width": "70%",
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholder="Password Confirmation"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
              "borderColor": "gray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#808080",
              "fontSize": 15,
              "fontWeight": "500",
              "height": 50,
              "paddingHorizontal": 10,
            },
            Object {
              "backgroundColor": "#fff",
              "borderColor": "gray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#808080",
              "flexDirection": "row",
              "fontSize": 15,
              "fontWeight": "500",
              "height": 50,
              "marginBottom": 10,
              "marginVertical": 5,
              "paddingHorizontal": 10,
              "width": "70%",
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "center",
            "backgroundColor": "#157dfb",
            "borderRadius": 5,
            "height": 50,
            "justifyContent": "center",
            "marginTop": 10,
            "marginVertical": 3,
            "opacity": 1,
            "width": 200,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#fff",
                "fontSize": 20,
                "fontWeight": "700",
              },
              undefined,
            ]
          }
        >
          Sign Up
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 20,
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontWeight": "300",
          },
          Object {
            "fontSize": 16,
            "textAlign": "center",
            "width": "60%",
          },
        ]
      }
    >
      If you have an account
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontWeight": "300",
              },
              Object {
                "color": "#157dfb",
                "fontSize": 16,
                "fontWeight": "500",
                "textAlign": "center",
              },
            ]
          }
        >
          please login here!
        </Text>
      </View>
    </Text>
  </View>
</View>
`;

exports[`Rendering should match to snapshot - login 1`] = `
<View
  style={
    Object {
      "height": 1234,
    }
  }
>
  <View
    keyboardShouldPersistTaps={true}
    onLayout={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <Modal
      animationType="none"
      hardwareAccelerated={false}
      transparent={true}
      visible={false}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#00000040",
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "space-around",
            "zIndex": 0,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "borderRadius": 10,
              "display": "flex",
              "height": 100,
              "justifyContent": "space-around",
              "width": 100,
              "zIndex": 0,
            }
          }
        >
          <ActivityIndicator
            animating={false}
            color="#157dfb"
            hidesWhenStopped={true}
            size="large"
          />
        </View>
      </View>
    </Modal>
    <Text
      style={
        Array [
          Object {
            "fontWeight": "300",
          },
          Array [
            Object {
              "color": "#808080",
              "fontSize": 18,
              "paddingBottom": 35,
              "paddingTop": 15,
            },
            Object {
              "paddingBottom": 20,
              "paddingHorizontal": "15%",
              "textAlign": "center",
            },
          ],
        ]
      }
    >
      Please enter your credentials in order to login
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "width": "100%",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholder="username"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
              "borderColor": "gray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#808080",
              "fontSize": 15,
              "fontWeight": "500",
              "height": 50,
              "paddingHorizontal": 10,
            },
            Object {
              "backgroundColor": "#fff",
              "borderColor": "gray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#808080",
              "flexDirection": "row",
              "fontSize": 15,
              "fontWeight": "500",
              "height": 50,
              "marginBottom": 10,
              "marginVertical": 5,
              "paddingHorizontal": 10,
              "width": "70%",
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholder="password"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
              "borderColor": "gray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#808080",
              "fontSize": 15,
              "fontWeight": "500",
              "height": 50,
              "paddingHorizontal": 10,
            },
            Object {
              "backgroundColor": "#fff",
              "borderColor": "gray",
              "borderRadius": 5,
              "borderWidth": 1,
              "color": "#808080",
              "flexDirection": "row",
              "fontSize": 15,
              "fontWeight": "500",
              "height": 50,
              "marginBottom": 10,
              "marginVertical": 5,
              "paddingHorizontal": 10,
              "width": "70%",
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "center",
            "backgroundColor": "#157dfb",
            "borderRadius": 5,
            "height": 50,
            "justifyContent": "center",
            "marginTop": 10,
            "marginVertical": 3,
            "opacity": 1,
            "width": 200,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#fff",
                "fontSize": 20,
                "fontWeight": "700",
              },
              undefined,
            ]
          }
        >
          Login
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 20,
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontWeight": "300",
          },
          Object {
            "fontSize": 16,
            "textAlign": "center",
            "width": "60%",
          },
        ]
      }
    >
      If you do not have an account
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontWeight": "300",
              },
              Object {
                "color": "#157dfb",
                "fontSize": 16,
                "fontWeight": "500",
                "textAlign": "center",
              },
            ]
          }
        >
          please create one here!
        </Text>
      </View>
    </Text>
  </View>
</View>
`;
